Add support for `Kokkos::resize` in the forward mode

gojakuch opened the pull request:
This PR extends the Kokkos support in Clad by providing custom pushforwards for different overloads of the `Kokkos::resize` function.

vgvassilev commented:
@kliegeois, @brian-kelley can you take a look? Looks good to me.

kliegeois commented on unittests/Kokkos/ViewBasics.cpp:
```diff
@@ -160,4 +160,48 @@ TEST(ViewBasics, TestDeepCopy2Reverse) {
+  Kokkos::View<double** [3], Kokkos::LayoutLeft, Kokkos::HostSpace> a("a", 3,
+                                                                      2);
+
+  Kokkos::deep_copy(a, 3 * x + y);
+  a(2, 1, 0) = x * y;
+
+  Kokkos::resize(Kokkos::WithoutInitializing, a, 5, 5);
+
+  return a(2, 1, 0);
```
I would suggest here (and on the following test as well) to change a little bit the test to return a value of the view that has been allocated during the resize.

For example doing something like this:
```
  Kokkos::deep_copy(a, 3 * x + y);

  Kokkos::resize(Kokkos::WithoutInitializing, a, 5, 5);

  a(4, 4, 0) = x * y;

  return a(4, 4, 0);
```

gojakuch replied:
oh sure, I can add these tests, that seems useful, but my initial goal with the tests I wrote was to check if the existing values are preserved after the resize, so I'd keep them

kliegeois commented on include/clad/Differentiator/KokkosBuiltins.h:
```diff
@@ -39,6 +39,29 @@ inline void deep_copy_pushforward(const View1& dst, const View2& src, T param,
+resize_pushforward(View& v, const size_t n0, const size_t n1, const size_t n2,
+                   const size_t n3, const size_t n4, const size_t n5,
+                   const size_t n6, const size_t n7, View& d_v,
+                   const size_t d_n0, const size_t d_n1, const size_t d_n2,
+                   const size_t d_n3, const size_t d_n4, const size_t d_n5,
+                   const size_t d_n6, const size_t d_n7) {
```
All these variables are not used in the function body.

I am almost sure that Clad requires to provide pushforward functions that takes as input the primal variables and each one of the derivatives but some code that use that function and that build with warning as errors will not compile due to unused variables.

If I recall correctly, a way to avoid that would be to have either:


```
(void) d_n0; 
(void) d_n1; 
(void) d_n2;
(void) d_n3; 
(void) d_n4; 
(void) d_n5;
(void) d_n6; 
(void) d_n7;
```
in the body of the function

or

replace 
```
                   const size_t d_n0, const size_t d_n1, const size_t d_n2,
                   const size_t d_n3, const size_t d_n4, const size_t d_n5,
                   const size_t d_n6, const size_t d_n7) {
```
with
```
                   const size_t /*d_n0*/, const size_t /*d_n1*/, const size_t /*d_n2*/,
                   const size_t /*d_n3*/, const size_t /*d_n4*/, const size_t /*d_n5*/,
                   const size_t /*d_n6*/, const size_t /*d_n7*/) {
```

gojakuch replied:
sure, I'll address that if it can be a problem potentially.

kliegeois commented on include/clad/Differentiator/KokkosBuiltins.h:
```diff
@@ -39,6 +39,29 @@ inline void deep_copy_pushforward(const View1& dst, const View2& src, T param,
+inline void resize_pushforward(
+    const I& arg, View& v, const size_t n0, const size_t n1, const size_t n2,
+    const size_t n3, const size_t n4, const size_t n5, const size_t n6,
+    const size_t n7, const dI& d_arg, View& d_v, const size_t d_n0,
+    const size_t d_n1, const size_t d_n2, const size_t d_n3, const size_t d_n4,
+    const size_t d_n5, const size_t d_n6, const size_t d_n7) {
```
Same as above.

kliegeois commented on unittests/Kokkos/ViewBasics.cpp:
```diff
@@ -160,4 +160,48 @@ TEST(ViewBasics, TestDeepCopy2Reverse) {
+  Kokkos::View<double** [3], Kokkos::LayoutLeft, Kokkos::HostSpace> a("a", 3,
+                                                                      2);
+
+  Kokkos::deep_copy(a, 3 * x + y);
+  a(2, 1, 0) = x * y;
+
+  Kokkos::resize(a, 5, 5);
+
+  return a(2, 1, 0);
```
Same as above.

kliegeois reviewed (commented):
Overall it looks good to me.
I just have couple of quick comments.

kliegeois commented on include/clad/Differentiator/KokkosBuiltins.h:
```diff
@@ -40,24 +40,28 @@ inline void deep_copy_pushforward(const View1& dst, const View2& src, T param,
+                               const size_t n1, const size_t n2,
+                               const size_t n3, const size_t n4,
+                               const size_t n5, const size_t n6,
+                               const size_t n7, const dI& d_arg, View& d_v,
```
Same for `d_arg` 😃

kliegeois reviewed (commented):
Thanks for the modifications!

Just one extra comment.

gojakuch commented:
@vgvassilev @kliegeois I think I've addressed all the suggestions and we can merge it. I put everything in separate commits, we can squash that obviously.

vgvassilev commented:
@kliegeois, if that looks good to you we can merge it.

kliegeois commented:
It looks good to me thanks!
